File: common/ParserTool.go
package common

import (
	"errors"
	"image/color"
	"strconv"
	"strings"
	"time"
)

func Parser_int32(content string, attachmentValue *int32) error {
	result, err := strconv.ParseInt(content, 10, 32)
	if err != nil {
		return err
	}
	*attachmentValue = int32(result)
	return nil
}
func Parser_int64(content string, attachmentValue *int64) error {
	result, err := strconv.ParseInt(content, 10, 64)
	if err != nil {
		return err
	}
	*attachmentValue = result
	return nil
}
func Parser_float32(content string, attachmentValue *float32) error {
	result, err := strconv.ParseFloat(content, 32)
	if err != nil {
		return err
	}
	*attachmentValue = float32(result)
	return nil
}
func Parser_float64(content string, attachmentValue *float64) error {
	result, err := strconv.ParseFloat(content, 64)
	if err != nil {
		return err
	}
	*attachmentValue = result
	return nil
}
func Parser_bool(content string, attachmentValue *bool) error {
	if content == "true" {
		*attachmentValue = true
	} else if content == "false" {
		*attachmentValue = true
	} else {
		return errors.New("can't parser bool " + content)
	}
	return nil
}
func Parser_string(content string, attachmentValue *string) error {

	*attachmentValue = content
	return nil
}
func Parser_color(contentStr string, attachmentValue *color.RGBA) error {
	content := strings.TrimPrefix(contentStr, "#")
	contentLen := strings.Count(content, "") - 1
	if contentLen != 6 && contentLen != 8 { // (r + g + b) * 2
		return errors.New("error on parser color")
	}

	color64, err := strconv.ParseInt(content, 16, 64) //字串到数据整型
	if err != nil {
		return err
	}
	if contentLen == 8 {
		red := color64 >> 24
		green := (color64 & 0x00FF0000) >> 16
		blue := (color64 & 0x0000FF00) >> 8
		alpha := color64 & 0x000000FF
		*attachmentValue = color.RGBA{uint8(red), uint8(green), uint8(blue), uint8(alpha)}
	} else {
		red := color64 >> 16
		green := (color64 & 0x00FF00) >> 8
		blue := color64 & 0x0000FF
		*attachmentValue = color.RGBA{uint8(red), uint8(green), uint8(blue), 0}
	}

	return nil
}
func Parser_dateTime(content string, attachmentValue *time.Time) error {
	parse_str_time, err := time.Parse("2006-01-02 15:04:05", content)
	if nil != err {
		return err
	}
	*attachmentValue = parse_str_time
	return nil
}
func CheckValueLimit_dateTime(value time.Time, limitMin string, limitMax string) error {
	var min time.Time
	var max time.Time
	if limitMin != "" {
		err := Parser_dateTime(limitMin, &min)
		if nil != err {
			return err
		}

		if value.Before(min) {
			return errors.New("error on check limit ")
		}
	}
	if limitMax != "" {
		err := Parser_dateTime(limitMax, &max)
		if nil != err {
			return err
		}
		if value.After(max) {
			return errors.New("error on check limit ")
		}
	}
	return nil
}
func CheckValueLimit_int32(value int32, limitMin string, limitMax string) error {
	var min int32 = 0
	var max int32 = 0
	if limitMin != "" {
		err := Parser_int32(limitMin, &min)
		if nil != err {
			return err
		}

		if value < min {
			return errors.New("error on check limit ")
		}
	}
	if limitMax != "" {
		err := Parser_int32(limitMax, &max)
		if nil != err {
			return err
		}
		if value > max {
			return errors.New("error on check limit ")
		}
	}
	return nil
}
func CheckValueLimit_int64(value int64, limitMin string, limitMax string) error {
	var min int64 = 0
	var max int64 = 0
	if limitMin != "" {
		err := Parser_int64(limitMin, &min)
		if nil != err {
			return err
		}

		if value < min {
			return errors.New("error on check limit ")
		}
	}
	if limitMax != "" {
		err := Parser_int64(limitMax, &max)
		if nil != err {
			return err
		}
		if value > max {
			return errors.New("error on check limit ")
		}
	}
	return nil
}
func CheckValueLimit_float32(value float32, limitMin string, limitMax string) error {
	var min float32 = 0
	var max float32 = 0
	if limitMin != "" {
		err := Parser_float32(limitMin, &min)
		if nil != err {
			return err
		}

		if value < min {
			return errors.New("error on check limit ")
		}
	}
	if limitMax != "" {
		err := Parser_float32(limitMax, &max)
		if nil != err {
			return err
		}
		if value > max {
			return errors.New("error on check limit ")
		}
	}
	return nil
}
func CheckValueLimit_float64(value float64, limitMin string, limitMax string) error {
	var min float64 = 0
	var max float64 = 0
	if limitMin != "" {
		err := Parser_float64(limitMin, &min)
		if nil != err {
			return err
		}

		if value < min {
			return errors.New("error on check limit ")
		}
	}
	if limitMax != "" {
		err := Parser_float64(limitMax, &max)
		if nil != err {
			return err
		}
		if value > max {
			return errors.New("error on check limit ")
		}
	}
	return nil
}
func IsTypeCanCheckLimit(fieldType string) bool {
	if fieldType == "int32" ||
		fieldType == "int64" ||
		fieldType == "float32" ||
		fieldType == "float64" {
		return true
	}
	return false
}

File: runtime/csharpRuntime/csharpRuntime/core/ParserTool.cs

using System;
using System.Security.Policy;

namespace ConfigProto.core
{
    public class ParserTool
    {
        public static int Parser_Int32(string content, out string errorMsg)
        {
            errorMsg = "";
            int res = 0;
            if (!int.TryParse(content, out res))
            {
                errorMsg = "error on parser cell " + content + " must be int32 ";
                return 0;
            }
            return res;
        }
        public static long Parser_Int64(string content, out string errorMsg)
        {
            errorMsg = "";
            long res = 0;
            if (!long.TryParse(content, out res))
            {
                errorMsg = "error on parser cell " + content + " must be int64 ";
                return 0;
            }
            return res;
        }
        public static float Parser_Float32(string content, out string errorMsg)
        {
            errorMsg = "";
            float res = 0;
            if (!float.TryParse(content, out res))
            {
                errorMsg = "error on parser cell " + content + " must be float32 ";
                return 0;
            }
            return res;
        }
        public static double Parser_Float64(string content, out string errorMsg)
        {
            errorMsg = "";
            double res = 0;
            if (!double.TryParse(content, out res))
            {
                errorMsg = "error on parser cell " + content + " must be float64 ";
                return 0;
            }
            return res;
        }
        public static bool Parser_Bool(string content, out string errorMsg)
        {
            errorMsg = "";
            if (content == "true")
            {
                return true;
            }
            else if (content == "false")
            {
                return false;
            }
            errorMsg =  "error on parser cell " + content + " must be true or false ";
            return false;
        }
        public static string Parser_String(string content, out string errorMsg)
        {
            errorMsg = "";
            return content;
        }
        public static Color Parser_Color(string content, out string errorMsg)
        {
            errorMsg = "";
            if (content.StartsWith("#"))
            {
                content = content.Substring(1);
            }
            if (content.Length != 6 && content.Length != 8)
            {
                errorMsg = "error on parser color";
                return null;
            }
            long color64 = Parser_Int64(content, out errorMsg);
            if("" != errorMsg)
            {
                errorMsg = "error on parser color";
                return null;
            }
            if (content.Length == 8)
            {
                byte red = (byte) (color64 >> 24);
                byte green = (byte) ((color64 & 0x00FF0000) >> 16);
                byte blue = (byte) ((color64 & 0x0000FF00) >> 8);
                byte alpha = (byte) (color64 & 0x000000FF);
                return new Color(red, green, blue, alpha);
            }
            else
            {
                byte red = (byte) (color64 >> 16);
                byte green = (byte) ((color64 & 0x00FF00) >> 8);
                byte blue = (byte) (color64 & 0x0000FF);
                return new Color(red, green, blue);
            }
            return null;
        }
        public static DateTime Parser_DateTime(string content, out string errorMsg)
        {
            errorMsg = "";
            DateTime dateValue;
            if (!DateTime.TryParse(content, out dateValue))
            {
                errorMsg = "error on parser cell " + content + " must be  datetime ";
                return dateValue;
            }
            return dateValue;
        }
        public static void CheckValueLimit_Int32(int value, string limitMin, string limitMax, out string errorMsg)
        {
            errorMsg = "";

            int min;
            int max;
            if (limitMin != "")
            {
                min = Parser_Int32(limitMin, out errorMsg);
                if (errorMsg != ""){return;}
                if (value < min)
                {
                    errorMsg = "error on check limit ";
                    return;
                }
            }
            if (limitMax != "")
            {
                max = Parser_Int32(limitMax, out errorMsg);
                if (errorMsg != ""){return;}
                if (value > max)
                {
                    errorMsg = "error on check limit ";
                }
            }
            return;
        }
        public static void CheckValueLimit_Int64(long value, string limitMin, string limitMax, out string errorMsg)
        {
            errorMsg = "";

            long min;
            long max;
            if (limitMin != "")
            {
                min = Parser_Int64(limitMin, out errorMsg);
                if (errorMsg != ""){return;}
                if (value < min)
                {
                    errorMsg = "error on check limit ";
                    return;
                }
            }
            if (limitMax != "")
            {
                max = Parser_Int64(limitMax, out errorMsg);
                if (errorMsg != ""){return;}
                if (value > max)
                {
                    errorMsg = "error on check limit ";
                }
            }
            return;
        }
        public static void CheckValueLimit_Float32(float value, string limitMin, string limitMax, out string errorMsg)
        {
            errorMsg = "";

            float min;
            float max;
            if (limitMin != "")
            {
                min = Parser_Float32(limitMin, out errorMsg);
                if (errorMsg != ""){return;}
                if (value < min)
                {
                    errorMsg = "error on check limit ";
                    return;
                }
            }
            if (limitMax != "")
            {
                max = Parser_Float32(limitMax, out errorMsg);
                if (errorMsg != ""){return;}
                if (value > max)
                {
                    errorMsg = "error on check limit ";
                }
            }
            return;
        }
        public static void CheckValueLimit_Float64(double value, string limitMin, string limitMax, out string errorMsg)
        {
            errorMsg = "";

            double min;
            double max;
            if (limitMin != "")
            {
                min = Parser_Float64(limitMin, out errorMsg);
                if (errorMsg != ""){return;}
                if (value < min)
                {
                    errorMsg = "error on check limit ";
                    return;
                }
            }
            if (limitMax != "")
            {
                max = Parser_Float64(limitMax, out errorMsg);
                if (errorMsg != ""){return;}
                if (value > max)
                {
                    errorMsg = "error on check limit ";
                }
            }
            return;
        }
    }
}
